Core: Add a firewall framework + stm32 use case

GseoC opened the pull request:
The need for such framework arises from the fact that there are multiple hardware firewalls implemented across multiple products.  When it comes to firewalls, the purpose mostly stays the same: protect and filter hardware resources. 

Add a firewall framework that offers:
- Firewall device consumers APIs to check, acquire accesses against firewall controllers and firewall reconfiguration as there are cases when a firewall configuration could be dynamic.
- Firewall controllers APIs to register/unregister to the framework and populate their bus.

This firewall framework relies on the _access-controllers_ device tree binding that was acked in the Linux kernel and is present in the Linux-next branch.
How the arguments passed along with this property and how they're used by the controllers to filter the accesses or set configurations is platform-specific AND specific to the firewall controller. Meaning that all the filtering logic should be present in the firewall controller drivers. Firewall consumers do not have the knowledge on the controller and should rely on the device API.

As example, I have reworked and cabled the STM32 ETZPC driver (STM32MP1x series) onto the framework as it implements all of the firewall controllers ops. I also have the RIFSC driver (STM32MP2x series) ready but did not put it in this P-R to avoid pollution. The STM32 ETZPC device is responsible for filtering accesses based on security level, or co-processor isolation for any resource connected to it.

GseoC commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -0,0 +1,737 @@
+
+struct dt_id_attr {
+	/* The effective size of the array is meaningless here */
+	fdt32_t id_attr[1];
```
I'll remove this, it is useless and provokes a warning.

etienne-lms commented on core/include/drivers/firewall/firewall.h:
```diff
@@ -0,0 +1,113 @@
+#include <drivers/firewall/firewall_device.h>
+#include <stddef.h>
+#include <tee_api_defines.h>
+#include <types_ext.h>
```
`#include <mm/core_memprot.h>`
`#include <stdbool.h>

etienne-lms commented on core/include/drivers/firewall/firewall.h:
```diff
@@ -0,0 +1,113 @@
+#include <tee_api_defines.h>
+#include <types_ext.h>
+
+struct firewall_controller_ops;
```
Not used. I think this ops struct could be defined, and referred from `struct firewall_controller`. I would allows driver to use a `static const struct firewall_controller_ops` reference.

etienne-lms commented on core/include/drivers/firewall/firewall.h:
```diff
@@ -0,0 +1,113 @@
+ * @release_access:		Callback used to release resources taken by a
+ *				consumer when the access was acquired with
+ *				@acquire_access
+ * @check_memory_access:	Callback used to check access for a a consumer
```
s/a a/a/

etienne-lms reviewed (commented):
Comments for commit "core: firewall: add firewall framework".
I wonder if firewall.h and firewall_device.h really need a sub-directory: move them from _core/include/drivers/firewall/_ to _core/include/drivers/_

etienne-lms commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -334,6 +329,227 @@ static TEE_Result etzpc_pm(enum pm_op op, unsigned int pm_hint __unused,
+	     id == ETZPC_TZMA0_ID ? 0 : 1, tzma_base, tzma_base + prot_size);
+
+	if (paddr >= tzma_base && paddr < tzma_base + prot_size &&
+	    paddr + size - 1 < tzma_base + tzma_size)
```
```c
	if (core_is_buffer_inside(paddr, size, tzma_base, prot_size))

etienne-lms reviewed (commented):
Comments for "drivers: stm32_etzpc: new driver to use firewall API".

GseoC commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -402,7 +322,7 @@ static TEE_Result init_stm32mp1_drivers(void)
 }
 
-service_init_late(init_stm32mp1_drivers);
+driver_init_late(init_stm32mp1_drivers);
```
Removed that change

GseoC commented:
Updated with comments addressed.

I added a [modifications] commit that changes a bit the filtering policy of the ETZPC, please have a look at it as well @etienne-lms

GseoC commented on core/include/drivers/firewall.h:
```diff
@@ -0,0 +1,123 @@
+ *				services
+ *
+ * @ops:			Operation handlers
+ * @name:			Name of the firewall controller
```
Ok

GseoC commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -0,0 +1,707 @@
+	etzpc_unlock(etzpc_dev, exceptions);
+}
+
+static bool is_tzma_locked(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
```
Changing decprot as well then.

It really depends on the person IMHO but ok.

GseoC commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -0,0 +1,707 @@
+	unsigned int lock;
+};
+
+#if TRACE_LEVEL >= TRACE_DEBUG
```
mmm ok

GseoC commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -0,0 +1,707 @@
+ * The driver API is defined in header file stm32_etzpc.h.
+ *
+ * Driver registers a PM callback for restoration of the access permissions
+ * when it resumes.
```
I'm ok to remove it

GseoC commented on core/drivers/firewall/stm32_etzpc.c:
```diff
@@ -0,0 +1,707 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2015-2021, ARM Limited and Contributors. All rights reserved.
```
Indeed, thanks

GseoC commented:
Updated with comments addressed and squashed.

@jenswi-linaro / @jforissier do you want to have a look as well on the generic part?

GseoC commented:
Fixed a `__maybe_unused` attribute that was misplaced

etienne-lms commented:
In order to ease reviewers commenting the generic part, maybe it would be worth creating another P-R, maybe [RFC] prefixed, with only the 1 commit of your series, knowing that this current P-R illustrate how an access-controller/firewall device integrates with this new framework.

GseoC commented:
> In order to ease reviewers commenting the generic part, maybe it would be worth creating another P-R, maybe [RFC] prefixed, with only the 1 commit of your series, knowing that this current P-R illustrate how an access-controller/firewall device integrates with this new framework.

Fine, I'll do that
